composer/trainer/dist_strategy.py

```diff
@@ -437,36 +439,56 @@ def _param_init_fn(module: torch.nn.Module) -> None:
                         'This leaves parameters without initialization. Please add a ``param_init_fn`` or ``reset_parameters`` '
                         f'to module `{obj_name}`.')
 
-            # Choose which modules to FSDP wrap according to the following priority:
-            # If module has attribute `module._fsdp_wrap = ...`, always respect it
-            # Otherwise wrap if root object `obj.fsdp_wrap_fn(module)` is true.
-            def __auto_wrap_policy(module: torch.nn.Module, recurse: bool, nonwrapped_numel: int) -> bool:
-                if recurse:
-                    return True
-                should_be_wrapped = False
-                if hasattr(module, '_fsdp_wrap'):
-                    should_be_wrapped = bool(module._fsdp_wrap)
-                elif hasattr(obj, 'fsdp_wrap_fn') and isinstance(obj.fsdp_wrap_fn, Callable):
-                    should_be_wrapped = obj.fsdp_wrap_fn(module)
-
-                if should_be_wrapped and auto_microbatching:
-                    module.register_forward_hook(sync_hook)
-                    module.register_full_backward_hook(sync_hook)
-                return should_be_wrapped
-
-            if is_torch_2_0:
-
-                def _auto_wrap_policy_new(module: torch.nn.Module, recurse: bool, nonwrapped_numel: int) -> bool:
-                    return __auto_wrap_policy(module, recurse, nonwrapped_numel)
-
-                _auto_wrap_policy = _auto_wrap_policy_new
-
+            if version.parse(torch.__version__) > version.parse('2.1.0.dev'):
+                # CustomPolicy is only supported in torch v2.1.0-rc1 or higher
+                from torch.distributed.fsdp.wrap import CustomPolicy
+
+                def lambda_fn(module: torch.nn.Module) -> Union[bool, dict]:
+                    ret = False
+                    if hasattr(module, '_fsdp_wrap'):
+                        ret = bool(module._fsdp_wrap)
+                    elif hasattr(obj, 'fsdp_wrap_fn') and isinstance(obj.fsdp_wrap_fn, Callable):
+                        ret = obj.fsdp_wrap_fn(module)
+                        from composer.trainer.mosaic_fsdp_utils import _set_custom_fsdp_module_kwargs
+                        if isinstance(ret, dict):
+                            ret = _set_custom_fsdp_module_kwargs(ret, process_group_cache)
+                    if ret and auto_microbatching:
+                        module.register_forward_hook(sync_hook)
+                        module.register_full_backward_hook(sync_hook)
+                    return ret
+
+                _auto_wrap_policy = CustomPolicy(lambda_fn)
             else:
+                # Choose which modules to FSDP wrap according to the following priority:
+                # If module has attribute `module._fsdp_wrap = ...`, always respect it
+                # Otherwise wrap if root object `obj.fsdp_wrap_fn(module)` is true.
+                def __auto_wrap_policy(module: torch.nn.Module, recurse: bool, nonwrapped_numel: int) -> bool:
```

Comment:
Is there a better name for this function that doesn't differentiate by one vs two underscores?

Reply:
This is an inner function that is rewrapped based on torch version. I agree it's not great...